InfiniBand: Add support of IP over InfiniBand

cathay4t opened the pull request:
Supporting IP over InfiniBand via:

```yml
- name: mlx5_ib1.8013
  type: infiniband
  state: up
  infiniband:
    base-iface: mlx5_ib1
    mode: datagram
    pkey: '0x8013'
```

Limitations:
 * Will raise NmstateValueError when setting IP over InfiniBand interface
   as bridge port. This is because they don't have ethernet layer.
   Please use RDMA over Converged Ethernet (RoCE).

 * Will raise NmstateValueError when setting IP over InfiniBand
   interface as bond port which is not in "active-backup" bond mode.
   This is Linux kernel limitation. Please use RDMA over Converged
   Ethernet for other bond modes.

To run the integration test case on real hardware of infiniband, please
use this command:

    sudo env TEST_REAL_NIC=mlx5_ib0 pytest-3 -vv  ./infiniband_test.py

Test cases are included.

dominikholler commented:
Thanks @cathay4t .
From oVirt point of view, it would be enough if something generic, independent from infiniband, would work, e.g.


```{'interfaces':
[
  {
      'name': 'ib0',
      'state': 'up',
      'mtu': 1500,
      'ipv4': {
          'enabled': True,
          'address': [{'ip': '1.2.3.4', 'prefix-length': 24}],
          'dhcp': False
          },
          'ipv6': {'enabled': False}
   },
      'routes': {
          'config': [
          {
              'next-hop-address': '1.2.3.1',
              'next-hop-interface': 'eth1',
              'destination': '0.0.0.0/0',
              'state': 'absent',
              'table-id': 0
          }
        ]
     }
  ]
}

```

or https://bugzilla.redhat.com/show_bug.cgi?id=1856279#c4 , just like Ethernet without Linux bridges.

dominikholler commented on libnmstate/ifaces/infiniband.py:
```diff
@@ -0,0 +1,94 @@
+
+    def _validate_mandatory_properties(self):
+        if self.is_up:
+            if InfiniBand.MODE not in self._ib_config:
```
Why is mode required, if not changed?

cathay4t replied:
When not changed, the mode will copy from running state.
This only validate when user try to create new ib pkey interface.

dominikholler replied:
I am not yet convinced: https://pastebin.com/cc3LVk7M

dominikholler replied:
The scenario works for managed interfaces, so I am fine.

cathay4t commented:
@dominikholler Thanks for feedback.

I have included test cases for IP over InfiniBand interface to:
 * Change IP address.
 * Add to Bond port.

The IP over InfiniBand interface cannot be a bridge port as they don't have ethernet layer.
The IP over InfiniBand interface can only be bond port when bond is in `active-backup` mode which is limited by linux kernel.

The RoCE(RDMA over Converged Ethernet) does not have above restriction. The nmstate will treat the RoCE interfaces as normal ethernet interface, no code need to be amended.

cathay4t commented:
@ffmancera Please review this PR ASAP as I only borrowed the infiniband servers for a very short time.

cathay4t commented:
The infiniband integration test cases has been ran on Mellanox MT27700 in datagram mode for 10 times. No failure.

ffmancera commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -26,11 +26,13 @@
 from libnmstate.schema import Interface
 from libnmstate.schema import InterfaceType
 from libnmstate.schema import InterfaceState
+from libnmstate.schema import BondMode
```
Could you please sort this in alphabetical order? Thanks.

ffmancera commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -197,6 +201,52 @@ def _validate_vlan_mtu(self):
+                for port_name in iface.port:
+                    port_iface = self._ifaces.get(port_name)
+                    if (
+                        port_iface
```
This is validated on `_validate_unknown_port`. IMO, this is not necessary.

ffmancera commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -197,6 +201,52 @@ def _validate_vlan_mtu(self):
+                for port_name in iface.port:
+                    port_iface = self._ifaces.get(port_name)
+                    if (
+                        port_iface
```
Same here.

ffmancera commented on libnmstate/nm/plugin.py:
```diff
@@ -136,6 +137,7 @@ def get_interfaces(self):
             iface_info.update(nm_user.get_info(self.context, dev))
             iface_info.update(nm_lldp.get_info(self.client, dev))
             iface_info.update(nm_team.get_info(dev))
+            iface_info.update(get_infiniband_info(applied_config))
```
Do we want to import the functions now? If so, I would refactor the others `nm_...get_info()` in another PR, I also prefer it.

cathay4t replied:
I prefer function over import the whole module.

ffmancera commented on libnmstate/nm/profile.py:
```diff
@@ -417,11 +418,12 @@ def _generate_connection_settings(self, ifname):
-        )
-        if wired_setting:
-            settings.append(wired_setting)
+        if self.iface.type != InterfaceType.INFINIBAND:
+            wired_setting = wired.create_setting(
+                original_state_wired, self._remote_conn
+            )
+            if wired_setting:
+                settings.append(wired_setting)
```
I do not understand why is this needed. Could you explain it?

cathay4t replied:
Included these comments:
  
            # The IP over InfiniBand has its own setting for MTU and does not
            # have ethernet layer.

ffmancera commented on libnmstate/nm/profile.py:
```diff
@@ -456,6 +458,12 @@ def _generate_connection_settings(self, ifname):
             )
             settings.extend(nm_ovs.create_interface_setting(patch_state))
+        elif self.iface.type == InterfaceType.INFINIBAND:
+            ib_setting = create_infiniband_setting(
```
Same here. I prefer importing only the function. Will refactor the others in another PR.

cathay4t replied:
Yes. I prefer import function only.

ffmancera commented on libnmstate/nm/profile.py:
```diff
@@ -570,6 +578,9 @@ def classify_profile_for_actions(self, groups):
                     is_absent
                     and self.nmdev.is_software()
                     and self.nmdev.get_device_type() != NM.DeviceType.VETH
+                    # InfiniBand PKEY interface cannot be removed even
+                    # it is software based
+                    and self.iface.type != InterfaceType.INFINIBAND
```
Why not? Does it stay forever on the system? What if the user wants to remove it? Is it not possible?

cathay4t replied:
I haven't take a deep look on this yet. Just got failure when try to delete it.

Will update it here.

cathay4t replied:
Above line has been removed. Somehow pkey can be removed and failure found any more.

ffmancera replied:
Thanks!

ffmancera reviewed (changes requested):
Thank for all the work, I've made some comments. Thanks!

ffmancera reviewed (approved):
LGTM, thanks!

cathay4t committed:
InfiniBand: Add support of IP over InfiniBand

Supporting IP over InfiniBand via:

```yml
- name: mlx5_ib1.8013
  type: infiniband
  state: up
  infiniband:
    base-iface: mlx5_ib1
    mode: datagram
    pkey: '0x8013'
```

Limitations:
 * Will raise NmstateValueError when setting IP over InfiniBand interface
   as bridge port. This is because they don't have ethernet layer.
   Please use RDMA over Converged Ethernet (RoCE).

 * Will raise NmstateValueError when setting IP over InfiniBand
   interface as bond port which is not in "active-backup" bond mode.
   This is Linux kernel limitation. Please use RDMA over Converged
   Ethernet for other bond modes.

Due to bug of kernel bug(https://bugzilla.redhat.com/1883478), we are
using NetworkManager for query action.

To run the integration test case on real hardware of infiniband, please
use this command:

    sudo env TEST_REAL_NIC=mlx5_ib0 pytest-3 -vv  ./infiniband_test.py

Test cases are included.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
Timeout on waiting QE feedback. Will wok in separate PR when issue been found.
